### DISCUSSION TITLE
How do I create a primary key for a table dynamically if one doesn't exist in a meta class?

### DISCUSSION THREAD

**Discussion by swquinn** [sqlalchemy/sqlalchemy#7892] [Usage Questions] [answered]
In SQLAlchemy 1.3.24 and earlier I am able to create a primary key for a model dynamically, that is if a primary key was not defined on a model class, I can create one. To do this, in the `__init__` function of the custom metaclass I would check for the existence of a primary key and if the class didn't have one I would create a column with `primary_key` set to True and assign that to the `cls.id` property.

```python
class BaseModelMeta(DeclarativeMeta):
    def __init__(cls, classname, bases, dict_):
        has_primary = any(
            [False] + [value.primary_key for value in cls.__dict__.values() if isinstance(value, sa.Column)]
        )

        if not has_primary:
            cls.id = sa.Column("id", sa.Integer, nullable=False, primary_key=True)
            cls.id._creation_order = 1
        super().__init__(classname, bases, dict_)

...

Model = declarative_base(name="Model", metaclass=BaseModelMeta)

class Foo(Model):
    __tablename__ = "Foo"
```

The above correctly creates an `id` column for defined models where the primary key is not explicitly defined.

In SQLAlchemy 1.4.32, I receive the error: "Mapper mapped class Foo->foo could not assemble any primary key columns for mapped table 'foo'"

I suspect this should be possible and I'm just missing something, most likely doing something out of order or incorrectly. I've been digging into this for a little while now and can't quite figure it out. Any advice?

**Comment by swquinn** [comment]
After having an "ah ha" moment, just after asking this question I tried the following:

```python
class BaseModelMeta(DeclarativeMeta):
    def __init__(cls, classname, bases, dict_):
        has_primary = any(
            [False] + [value.primary_key for value in cls.__dict__.values() if isinstance(value, sa.Column)]
        )

        if not has_primary:
            dict_.update({ "id": sa.Column("id", sa.BigInteger, nullable=False, primary_key=True) })
            dict_["id"]._creation_order = 1
        super().__init__(classname, bases, dict_)
```

... and that appears to have worked. Is this the correct way to do what I'm looking to do in SQLAlchemy 1.4+?

**Reply by zzzeek** [reply]
Well, most simply, you can just have the "id" column on your base class and it will be used by default, the example at https://docs.sqlalchemy.org/en/14/orm/declarative_mixins.html#augmenting-the-base shows this.     I would use that if there isn't some reason that can't work.

Next, if you want something more on the fly, use an event handler, in this case a good choice is [instrument_class](https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=instrumentationevents#sqlalchemy.orm.MapperEvents.instrument_class), like this:

```py
from sqlalchemy import event

Base = declarative_base()


@event.listens_for(Base, "instrument_class", propagate=True)
def receive_class_instrument(mapper, cls):
    if not cls.__table__.primary_key:
        cls.__table__.append_column(
            Column("id", Integer, primary_key=True)
        )

class A(Base):
    __tablename__ = 'a'

    data = Column(String)

```

if using Python class creation mechanics, I also would normally advise using [__init_subclass__()](https://peps.python.org/pep-0487/) method which makes the vast majority of "metaclass" uses unnecessary.   However in this case it seems like declarative's metaclass is grabbing the class dictionary before such a hook can be used, which might be something I want to fix.

but anyway, I'd try to stay away from metaclasses as much as possible.   The mixin approach should be preferred and if not then go with the event handler.

**Reply by swquinn** [reply]
Thanks for the reply, @zzzeek !

Good to know about the event listener for `instrument_class` and the `init_subclass` method. The raison d'etre for this is born out of a preference (maybe misplaced) to allow for the autogeneration of an "id" column in the event one is not explicitly specified on a model. I fully understand that this is creating "magic" functionality and is almost definitely an ill-advised practice.

Still... in either the listener, or in the `init_subclass()` method, when appending the column, would doing something like this work to ensure the column ends up in the first position?

```python
id_column = Column("id", Integer, primary_key=True)
id_column._creation_order = 1
cls.__table__.append_column(id_column)
```

As an aside, I did notice in the more recent declarative style documentation that there appears to be a move towards mixins in general. In addition to what you write above, it sounds like there is a move away from user-defined metaclasses? Are there other useful methods or events that would be worthwhile knowing about, and is [MapperEvents the best place to find the events, at least, documented](https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=instrument_class#mapper-events)?

If you have the time and would be willing to indulge one last question in this thread, I'd be curious to know from your perspective when one should reach for metaclasses?

**Reply by swquinn** [reply]
Using the `instrument_class` event worked, but I haven't been able to affect the creation order and place the `id` column as the first column in the schema. Is there a way to do this using the instrument class event?

However, I was struggling with using `__init_subclass__`. It didn't seem to have the necessary data structures to add a column to the table generated by the model.

**Reply by zzzeek** [reply]
> Thanks for the reply, @zzzeek !
> 
> Good to know about the event listener for `instrument_class` and the `init_subclass` method. The raison d'etre for this is born out of a preference (maybe misplaced) to allow for the autogeneration of an "id" column in the event one is not explicitly specified on a model. I fully understand that this is creating "magic" functionality and is almost definitely an ill-advised practice.

there's nothing wrong at all with the pattern, it's just best to use the hooks that are provided to do it.


> 
> Still... in either the listener, or in the `init_subclass()` method, when appending the column, would doing something like this work to ensure the column ends up in the first position?
> 
> ```python
> id_column = Column("id", Integer, primary_key=True)
> id_column._creation_order = 1
> cls.__table__.append_column(id_column)

there's not a good way to control the "position" of a column, and by the time append_column() is called, the order is set up, the new column is appended to the end.


> ```
> 
> As an aside, I did notice in the more recent declarative style documentation that there appears to be a move towards mixins in general. 

declarative has been very oriented around mixins for many years now.

> In addition to what you write above, it sounds like there is a move away from user-defined metaclasses? 

definitely because they are difficult to work with and make inheritance much more difficult.   we try to have enough hooks on declarative classes that custom metaclasses should not be needed, and of course `__init_subclass__()` should allow the remaining "metaclass" use cases to go away.


> Are there other useful methods or events that would be worthwhile knowing about, and is [MapperEvents the best place to find the events, at least, documented](https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=instrument_class#mapper-events)?

the events there are useful and then there's just the mixin patterns documented at https://docs.sqlalchemy.org/en/14/orm/declarative_mixins.html

> 
> If you have the time and would be willing to indulge one last question in this thread, I'd be curious to know from your perspective when one should reach for metaclasses?

the one thing metaclasses can do right now is allow one to intercept a class-level attribute being assigned, as well as they can be used for some very esoteric patterns being able to override special methods on classes.   but none of that belongs with regular mapped classes.

**Reply by zzzeek** [reply]
> Using the `instrument_class` event worked, but I haven't been able to affect the creation order and place the `id` column as the first column in the schema. Is there a way to do this using the instrument class event?

instrument class is called after declarative has already built the Table object where there is not currently a public API hook to control the order of columns.  

> 
> However, I was struggling with using `__init_subclass__`. It didn't seem to have the necessary data structures to add a column to the table generated by the model.

as I mentioned at https://github.com/sqlalchemy/sqlalchemy/discussions/7892#discussioncomment-2496450, this is not working right now.    I can have that working for 1.4.35 however, if you file an issue.

if you use the mixin approach, I think that will put the "id" column first.

**Reply by swquinn** [reply]
Thanks for the very thoughtful and detailed replies, @zzzeek ! I would be happy to file an issue for getting `__init_subclass__` implemented. I must have completely missed your comment about about it not working. Sorry for that! Is there anything in particular you would need in the issue for the `__init_subclass__` issue?

Thanks!

**Reply by zzzeek** [reply]
I've file the issue at #7900 and there's some code in review.

**Reply by zzzeek** [reply]
I've also set up #7910, which is more the real problem here, so you can insert a PK column at the front of the Table.

**Comment by swquinn** [comment, marked as answer]
See the responds from @zzzeek here: https://github.com/sqlalchemy/sqlalchemy/discussions/7892#discussioncomment-2496450 for the answer to this question.

Thanks again, @zzzeek !
